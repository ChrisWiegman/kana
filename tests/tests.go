package tests

import (
	"bytes"
	"os"
	"os/exec"
	"path/filepath"
	"testing"

	"github.com/ChrisWiegman/kana/internal/helpers"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/mitchellh/go-homedir"
)

type Test struct {
	Description string
	Command     []string
	Docker      bool
}

func Setup() {
	testDirectory := "./kana"

	dirExists, err := helpers.PathExists(testDirectory)
	if err != nil {
		panic(err)
	}

	if dirExists {
		err = os.RemoveAll(testDirectory)
		if err != nil {
			panic(err)
		}

		err = os.Mkdir(testDirectory, 0755) //nolint: mnd
		if err != nil {
			panic(err)
		}
	}

	err = os.Chdir(testDirectory)
	if err != nil {
		panic(err)
	}
}

func Teardown(docker bool) {
	home, err := homedir.Dir()
	if err != nil {
		panic(err)
	}

	appDirectory := filepath.Join(home, ".config", "kana")

	err = os.RemoveAll(appDirectory)
	if err != nil {
		panic(err)
	}

	if docker {
		commands := [][]string{
			{"kill", "$(docker ps -q)"},
			{"container", "prune", "-f"},
			{"network", "prune", "-f"},
			{"volume", "prune", "-f"},
			{"system", "prune", "-a", "-f"}}

		for _, command := range commands {
			cmd := exec.Command("docker", command...)
			err = cmd.Run()
			if err != nil {
				panic(err)
			}
		}
	}
}

func RunSnapshotTest(testCases []Test, t *testing.T) {
	for _, test := range testCases {
		t.Run(test.Description, func(t *testing.T) {
			cmd := exec.Command("../../build/kana", test.Command...) //nolint: gosec
			var out bytes.Buffer
			cmd.Stdout = &out
			err := cmd.Run()
			if err != nil {
				t.Fatalf("Unexpected error: %v", out.String())
			}

			snaps.MatchSnapshot(t, out.String())

			Teardown(test.Docker)
		})
	}
}
