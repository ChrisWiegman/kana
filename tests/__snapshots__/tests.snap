
[TestConfig/Test_the_default_config_command - 1]
┌──────────────────────┬─────────────────────┬─────────────────────┐
│       Setting        │    Global Value     │     Local Value     │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ activate             │ [1mtrue[0m                │ [1mtrue[0m                │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ admin.email          │ [1madmin@sites.kana.sh[0m │ [1madmin@sites.kana.sh[0m │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ admin.password       │ [1mpassword[0m            │ [1mpassword[0m            │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ admnin.username      │ [1madmin[0m               │ [1madmin[0m               │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ automaticLogin       │ [1mtrue[0m                │ [1mtrue[0m                │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ database             │ [1mmariadb[0m             │ [1mmariadb[0m             │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ databaseClient       │ [1mphpmyadmin[0m          │ [1mphpmyadmin[0m          │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ databaseVersion      │ [1m11[0m                  │ [1m11[0m                  │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ environment          │ [1mlocal[0m               │ [1mlocal[0m               │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ imageUpdateDays      │ [1m7[0m                   │                     │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ mailpit              │ [1mfalse[0m               │ [1mfalse[0m               │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ multisite            │ [1mnone[0m                │ [1mnone[0m                │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ php                  │ [1m8.2[0m                 │ [1m8.2[0m                 │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ plugins              │                     │                     │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ removeDefaultPlugins │ [1mfalse[0m               │ [1mfalse[0m               │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ ssl                  │ [1mfalse[0m               │ [1mfalse[0m               │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ scriptdebug          │ [1mfalse[0m               │ [1mfalse[0m               │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ theme                │                     │                     │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ type                 │ [1msite[0m                │ [1msite[0m                │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ wpdebug              │ [1mfalse[0m               │ [1mfalse[0m               │
├──────────────────────┼─────────────────────┼─────────────────────┤
│ xdebug               │ [1mfalse[0m               │ [1mfalse[0m               │
└──────────────────────┴─────────────────────┴─────────────────────┘

---

[TestConfig/Retrieve_the_PHP_value_from_the_config_command - 1]
8.2

---

[TestConfig/Test_the_config_command_with_json_output - 1]
{"Global":{"activate":true,"admin":{"email":"admin@sites.kana.sh","password":"password","username":"admin"},"automaticlogin":true,"database":"mariadb","databaseclient":"phpmyadmin","databaseversion":"11","environment":"local","imageupdatedays":7,"mailpit":false,"multisite":"none","php":"8.2","plugins":[],"removedefaultplugins":false,"scriptdebug":false,"ssl":false,"theme":"","type":"site","wpdebug":false,"xdebug":false},"Local":{"activate":true,"admin":{"email":"admin@sites.kana.sh","password":"password","username":"admin"},"automaticlogin":true,"database":"mariadb","databaseclient":"phpmyadmin","databaseversion":"11","environment":"local","mailpit":false,"multisite":"none","php":"8.2","plugins":null,"removedefaultplugins":false,"scriptdebug":false,"ssl":false,"theme":"","type":"site","wpdebug":false,"xdebug":false}}

---

[TestList/Test_the_default_list_command - 1]
┌──────┬──────┬─────────┐
│ Name │ Path │ Running │
└──────┴──────┴─────────┘

---

[TestList/Test_the_list_command_with_json_output - 1]
[]

---

[TestConfig/Retrieve_the_PHP_value_from_the_config_command_with_json_output - 1]
{"Setting":"php","Value":"8.2"}

---

[TestRoot/run_the_kana_root_command_without_further_input - 1]
Kana is a simple WordPress development tool designed for plugin and theme developers.

Usage:
  kana [command]

Available Commands:
  changelog   Open Kana's changelog in your browser
  config      View and edit the saved configuration for the app or the local site.
  db          Commands to easily import and export a WordPress database from an existing site
  destroy     Destroys the current WordPress site. This is a permanent change.
  export      Export the current config to a .kana.json file to save with your repo.
  flush       Flushes the cache and deletes all transients.
  help        Help about any command
  list        Lists all Kana sites and their associated status.
  open        Open the current site in your browser.
  start       Starts a new environment in the local folder.
  stop        Stops the WordPress development environment.
  version     Displays version information for the Kana CLI.
  wp          Run a wp-cli command against the current site.
  xdebug      Turns Xdebug on or off without having to stop and start the site.

Flags:
  -h, --help          help for kana
      --name string   Specify a name for the site, used to override using the current folder.
  -v, --verbose       Display debugging information along with detailed command output

Use "kana [command] --help" for more information about a command.

---

[TestRoot/Test_the_kana_root_command_with_json_output - 1]
Kana is a simple WordPress development tool designed for plugin and theme developers.

Usage:
  kana [command]

Available Commands:
  changelog   Open Kana's changelog in your browser
  config      View and edit the saved configuration for the app or the local site.
  db          Commands to easily import and export a WordPress database from an existing site
  destroy     Destroys the current WordPress site. This is a permanent change.
  export      Export the current config to a .kana.json file to save with your repo.
  flush       Flushes the cache and deletes all transients.
  help        Help about any command
  list        Lists all Kana sites and their associated status.
  open        Open the current site in your browser.
  start       Starts a new environment in the local folder.
  stop        Stops the WordPress development environment.
  version     Displays version information for the Kana CLI.
  wp          Run a wp-cli command against the current site.
  xdebug      Turns Xdebug on or off without having to stop and start the site.

Flags:
  -h, --help          help for kana
      --name string   Specify a name for the site, used to override using the current folder.
  -v, --verbose       Display debugging information along with detailed command output

Use "kana [command] --help" for more information about a command.

---

[TestVersion/Test_the_version_command_for_appropriate_output - 1]
Version: 1.0.0
Build Time: 2024-03-16_10:50:11PM

---

[TestVersion/Test_the_config_command_with_json_output - 1]
{"Version":"1.0.0","Timestamp":"2024-03-16_10:50:11PM"}

---
